docs/chapter-3/depth_buffer.md

```diff
@@ -253,6 +253,32 @@ Last thing we need is to add the depth attachment to the attachment list in the
 
 Instead of storing only the color attachment in pAttachments, we add the depth attachment there too.
 
+Now we have to adjust the renderpass synchronization. Previously, it was possible that multiple frames were rendered simultaneously by the GPU. This is a problem when using depth buffers, because one frame could overwrite the depth buffer while a previous frame is still rendering to it.
+
+We add a new subpass dependency on the `init_default_renderpass()` function that synchronizes accesses to depth attachments.
+
+```cpp
+VkSubpassDependency depth_dependency = {};
+depth_dependency.srcSubpass = VK_SUBPASS_EXTERNAL;
+depth_dependency.dstSubpass = 0;
+depth_dependency.srcStageMask = VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+depth_dependency.srcAccessMask = 0;
+depth_dependency.dstStageMask = VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+depth_dependency.dstAccessMask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
```

Comment:
I'm looking through the source code, couldn't this be combined with the other subpass dependency? Having it as a separate one is convenient for teaching but I'm curious if thats actually necessary.

Reply:
It’s not strictly necessary, you’re correct. As you also said, I think it’s nice to have them separated for clarity.

Reply:
Great, was just like "wait a minute..." and had to ask.

Reply:
Sorry to keep bothering you, but what is the point of having both early and late fragment tests here?

Reply:
You’re right, one of these is redundant in both the source and destination mask. My intent was to avoid people getting confused over different src/dst stage masks because the tutorial doesn’t mention that earlier/later stages are implicitly included for the src/dst mask respectively.

Reply:
From what I can see in <https://vulkan-tutorial.com/Depth_buffering#page_Render-pass>, it should be fine. Early fragment tests are always specified, together with attachment output (although one of the two stages is redundant here too :) ). Attachment output is logically later than late fragment tests, (and early fragment tests are logically earlier than late fragment tests), so late fragment tests reading from the depth buffer are implicitly included in both synchronization scopes.